### ISSUE TITLE
Strange behavior of Course Alarm

### ISSUE THREAD

**Issue by Corsair-63** [rgleason/watchdog_pi#62]
Windows 11 Pro ARM64 ver 22H2 22621.2215
OCPN 5.8.4-0+1637c28
Watchdog plugin 2.4.70.0
VDR plugin 1.3.28.0
Multiplexer - Shipmodul MiniPlex-3E-N2K - Firmware Version 3.18.2
GNSS - Digital Yacht GPS 160 at 38400 bps via multiplexer
Auto pilot / Compass - RayMarine via Sea Talk via Multiplexer
Wind vane - Nasa - Clipper-Target-Cruiser-V2.0 at 4800 via Multiplexer
Depth STW, Seawater temp, and some other data - Furuno GP1971F - Plotter/RADAR/Echo-sounder via N2K via multiplexer
AIS Rec / 2nd Echo sounder /2nd GNSS - at 38400 via multiplexer.

in second test mobile Pixel 5 app Bluetooth GPS output in BT COM port at 4800 bps.

also, when click on setup Course Alarm, never goes to the heading device reading, supposed to take first the COG figure. (when click on actual course)

first test:
- unable all inputs and only left the multiplexer data via UDP.
- no problem with GPS source
- exchanged to head device and gives course error.
- attached sequence and VDR file.
- WD_Course_GPS_XX, WD_Course_Compass_XX, VDR_Course.zip

second test:
- unable all inputs and only left BT COM port of mobile
- no problem with gps
- exchanged to head device and no success, but double check of input filters and found that not to be received HDM & HDG, removed these filter and alarm working fine.
- rotated mobile and alarm triggers.
- returning mobile to position and alarm reset.
- attached sequence and VDR file
- WD_Course_Compass_mobile_XX, GNSS_mobile, VDR_Course_mobile.zip.

third test:
- without restarting OCPN , unable all inputs and only left the multiplexer data via UDP. 
- alarm active and working.
- restarted OCPN and alarm not working. closed OCPN
- started setup software of multiplexer and unable convert HDG <-> HDT
- started OCPN and alarm WD course alarm not working
- right click on Course WD alarm setup, just click on ok , and after some seconds started working.
- when setup in multiplexer convert HDG <-> HDT unable to get alarm working in OCPN.
- in case of not converting course in multiplexer every time started OCPN the WD course alarm fails but exactly right click and click OK after some second come active and working.
- what else to say??

![WD_Course_GPS_02](https://github.com/rgleason/watchdog_pi/assets/76228638/90de433e-5284-4661-91cd-08e12694a9c9)
[VDR_Course.zip](https://github.com/rgleason/watchdog_pi/files/12536901/VDR_Course.zip)
[VDR_Course_mobile.zip](https://github.com/rgleason/watchdog_pi/files/12536902/VDR_Course_mobile.zip)
![GNSS_mobile](https://github.com/rgleason/watchdog_pi/assets/76228638/fa1c6056-c6dc-4771-bb67-e491a5c00cf1)
![WD_Course_Compass_01](https://github.com/rgleason/watchdog_pi/assets/76228638/d7becc91-6c14-40cd-a563-1d4e69152275)
![WD_Course_Compass_02](https://github.com/rgleason/watchdog_pi/assets/76228638/3f11bb25-2bf9-4367-baf8-ec301fe922d4)
![WD_Course_Compass_mobile_01](https://github.com/rgleason/watchdog_pi/assets/76228638/ca45c5aa-602a-4345-a22d-5861074b9f2d)
![WD_Course_Compass_mobile_02](https://github.com/rgleason/watchdog_pi/assets/76228638/0bb544bc-0b9b-4ff8-84a8-339d1e61f51f)
![WD_Course_Compass_mobile_03](https://github.com/rgleason/watchdog_pi/assets/76228638/022ddbbf-8b4d-40c4-a690-04f779eb898b)
![WD_Course_Compass_mobile_04](https://github.com/rgleason/watchdog_pi/assets/76228638/6fefa76f-606c-455b-89d0-73976252d4eb)
![WD_Course_GPS_01](https://github.com/rgleason/watchdog_pi/assets/76228638/b9493749-641a-4692-9268-8e894858a46b)

**Comment by Corsair-63** [comment]
Digging a little in above report, I've compared both compass sentences from both situations, and I have no clue about why mobile signal is immediately detected and not the multiplexer one?

> Input from Multiplexer (Mx)
> $HEHDT,287.0,T*22
> $HCHDG,287.00,,,0.04,E*20
> 
> Input from Mobile (Mb)
> $HCHDT,280.1,T*22
> $HCHDG,280.0,,,000.1,E*22
> 
> Mx $HEHDT,287.0,T*22
> Mb $HCHDT,280.1,T*22
> 
> Mx $HCHDG,280.0,,,000.1,E*22
> Mb $HCHDG,287.00,,,0.04,E*20
> 
> From NMEA talker ID:
> HC - Heading sensors: compass magnetic.
> HE - Heading sensors: gyro, north seeking.
> 
> HDT - Heading true.
> HDG - Heading, deviation and variation.
> 

only difference in sentences is one from multiplexer starting with HE and the other one by HC when mobile starts both with HC.
NMEA sentences in both inputs have the same arrangement conforming NMEA standard.

**Comment by Corsair-63** [comment]
Confirmed behaviour of WD Course alarm.
- run setup software of multiplexer and unable convert HDG <-> HDT
- started OCPN and WD off-course alarm still not active, in red and N/A
- after about 30 seconds converted in black figures and showing the difference with the setup course.
- also showing the arc of degrees in the boat for the course tolerance.
- attached screenshot of setup windows and running normal windows

so, it means that for any reason there is a conflict with $HC & $HE inputs because in the mobile sentences both are $HC and the WD course alarm is shown without delay.

also, pointing that when click on actual course the course read is COG from GPS and not from the Heading Device.

![WD_Course_Alarm_working](https://github.com/rgleason/watchdog_pi/assets/76228638/87b5db9c-27ef-4979-b5bc-75ee17aad1f0)
![WD_Course_Alarm_setup](https://github.com/rgleason/watchdog_pi/assets/76228638/73e87f1a-64ce-49a9-b28b-e037263960e0)

**Comment by rgleason** [comment]
Cosair, with this instrumentation:

1. GNSS - Digital Yacht GPS 160 at 38400 bps via multiplexer
2. Auto pilot / Compass - RayMarine via Sea Talk via Multiplexer
3. 2nd GNSS - at 38400 via multiplexer.

How are you filtering and prioritizing the nmea signals?
This is critical for Watchdog to be effective.

I have not yet tested your vdr files as we are working on a sizeable TP project, but I am following.
Your reports are excellent. Thank you.

**Comment by Corsair-63** [comment]
Everything doesn't need to be prioritized in OCPN because the work is done by Multiplexer and set on it as:
1. GNSS DY GPS 160
2. Autopilot
3. 2nd GNSS
also, bearing in mind that when not received same NMEA sentence is taken from the next in priority queue.
it means that only received one same sentence in OCPN in case of duplication by sources.
this help OCPN for not making the filtering sequence.

**Comment by rgleason** [comment]
Okay, have you captured the result that Watchdog sees in your VDR_course.zip above?
If so, what sentences do you see there? Perhaps I should look now.

**Comment by rgleason** [comment]
Well it doesn't look like the VDR_Course.txt is being filtered does it?
I think you'd better use the OpenCPN filter feature.

**Comment by Corsair-63** [comment]
> Well it doesn't look like the VDR_Course.txt is being filtered, does it?
VDR_Course.txt is with all data from all devices from Multiplexer and filtered by it, none duplicated sentence.

> What happens when you set Watchdog to Heading sensor?
with that log Course alarm comes to red and N/A.
that's with $HC & $HE from multiplexer (Sea Talk converted) being able convert HDG <-> HDT.

and if I unable the multiplexer input and activate the mobile one (VDR_Course_mobile.zip) the alarm is working correctly (that's the signal of GNSS, Compass & barometer of the mobile itself)

I haven't sent a log with convert HDG <-> HDT unable in Multiplexer but that works after some seconds of OCPN started and Off Course alarm works as expected.

**Comment by Corsair-63** [comment]
> Okay, have you captured the result that Watchdog sees in your VDR_course.zip above? If so, what sentences do you see there? Perhaps I should look now.

yes, with input GPS last picture (setup) and result the one with red figures and 90 degrees.

with Heading device, the 3rd picture (setup) and 4th the error red and N/A
all referenced pictures from original message.

**Comment by rgleason** [comment]
Corsair-63  This morning I tried both files and yes indeed there are some screwy things going on and I don't know why yet. But the fact is, garbage in garbage out.  You should look carefully at the nmea data stream being received by Watchdog. 

First of all I can confirm that the "Course" button in the Course Edit window does not work properly.

Just to be sure that watchdog course alarm was working properly using our tests I stripped down a Tactics nmea data file and ran it, and the course alarm works perfectly fine on either GPS or Heading buttons.

I attach it here as a compressed file so you can try it out without any instruments attached, just run VDR.
After you try this file, go back and compare it with your file perhaps.

[Tactics-sample2-1m.zip](https://github.com/rgleason/watchdog_pi/files/12550745/Tactics-sample2-1m.zip)

**Comment by Corsair-63** [comment]
Hi Rick,

I tried your file and works same than mine from Multiplexer.
since I removed in the Multiplexer the option of convert HDG <-> HDT, the alarm works as desired, only last some seconds to activate after start OCPN.

as you pointed, also, not available to set the course as heading device when click on "actual course" always COG.

filtering in this situation when supposed that all data came from the same stream, it's little complicated but I tried with the OCPN priority feature, there was quite a lot of entrances, so I erase all, and started all my connection to recreate, and I left as attached picture priority.jpg.

also, the priority as told before is done with multiplexer, and I attached picture of it, there are not so many output sentences as they are filtered on every cicle for not sending redundant from various sources.
![Priority](https://github.com/rgleason/watchdog_pi/assets/76228638/c4f9415f-4dca-4c45-8c5c-95b169a8bba1)
![Multiplexer](https://github.com/rgleason/watchdog_pi/assets/76228638/118dead1-25db-42eb-bfda-6d67b49ddbb5)
1. is autopilot.
2. is wind vane.
3. is GNSS DY
4. is AIS receiver that multiplex its own GPS, plus 2nd GNSS, plus second echo sounder.
N2K is Furuno Plotter with RADAR, triducer echo-sounder, GPS.

what I really don't know in priority window what are the NMEA 0183 virtual?

regarding the test with your log file, still not able to set the wind alarm nor true relative neither true absolute.

Carmelo
